<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>A birth-death model with heterogeneous fixed delays · DelaySSAToolkit.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link rel="canonical" href="https://palmtree2013.github.io/DelaySSAToolkit.jl/tutorials/heterogeneous_delay/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.044/juliamono.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">DelaySSAToolkit.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Tutorials</span><ul><li><a class="tocitem" href="../tutorials/">Tutorials</a></li><li><a class="tocitem" href="../bursty/">A bursty model with delay</a></li><li><a class="tocitem" href="../delay_degradation/">A birth-death example with delay degradation</a></li><li class="is-active"><a class="tocitem" href>A birth-death model with heterogeneous fixed delays</a><ul class="internal"><li><a class="tocitem" href="#Model-definition"><span>Model definition</span></a></li><li><a class="tocitem" href="#Heterogeneous-fixed-delays"><span>Heterogeneous fixed delays</span></a></li><li><a class="tocitem" href="#Visualisation"><span>Visualisation</span></a></li><li><a class="tocitem" href="#Distributed-delays"><span>Distributed delays</span></a></li><li><a class="tocitem" href="#Visualisation-2"><span>Visualisation</span></a></li><li><a class="tocitem" href="#References"><span>References</span></a></li></ul></li><li><a class="tocitem" href="../delay_oscillator/">An auto-regulatory model of oscillatory gene expression</a></li><li><a class="tocitem" href="../stochastic_delay/">A telegraph model of stochastic delay</a></li></ul></li><li><span class="tocitem">Algorithm</span><ul><li><a class="tocitem" href="../../algorithms/notations/">Notations and Basic Concepts</a></li><li><a class="tocitem" href="../../algorithms/delayrejection/">Delay Rejection Method Algorithm</a></li><li><a class="tocitem" href="../../algorithms/delaydirect/">Delay Direct Method Algorithm</a></li><li><a class="tocitem" href="../../algorithms/delaymnrm/">Delay Modified Next Reaction Method Algorithm</a></li></ul></li><li><a class="tocitem" href="../../theory/">Theory</a></li><li><a class="tocitem" href="../../api/">API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Tutorials</a></li><li class="is-active"><a href>A birth-death model with heterogeneous fixed delays</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>A birth-death model with heterogeneous fixed delays</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/palmtree2013/DelaySSAToolkit.jl/blob/master/docs/src/tutorials/heterogeneous_delay.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="A-birth-death-model-with-heterogeneous-fixed-delays"><a class="docs-heading-anchor" href="#A-birth-death-model-with-heterogeneous-fixed-delays">A birth-death model with heterogeneous fixed delays</a><a id="A-birth-death-model-with-heterogeneous-fixed-delays-1"></a><a class="docs-heading-anchor-permalink" href="#A-birth-death-model-with-heterogeneous-fixed-delays" title="Permalink"></a></h1><h2 id="Model-definition"><a class="docs-heading-anchor" href="#Model-definition">Model definition</a><a id="Model-definition-1"></a><a class="docs-heading-anchor-permalink" href="#Model-definition" title="Permalink"></a></h2><p>We study the following model with delayed production</p><p class="math-container">\[\emptyset \xrightarrow{A_n} X \text{ triggers } X \Rightarrow Y \text{ after $\tau$ time},\\
Y \xrightarrow{B_n} \emptyset.\]</p><p>This model is studied in [1]. This delayed birth–death process can be used to model the dynamics of chemical species, such as proteins, that are produced through a sequence of reactions. We first consider fixed birth delays, <span>$\tau_n$</span>, which are constant for a given cell but which vary across a population. It is assumed that across the population, production rates, <span>$A_n$</span>, degradation rates, <span>$B_n$</span>, and the fixed birth delays, <span>$\tau_n$</span>, follow a gamma distribution. A decrease in protein count is due to growth-induced dilution or enzymatic degradation, and is described by an instantaneous death process with rate <span>$B_n$</span>.</p><h3 id="Markovian-part"><a class="docs-heading-anchor" href="#Markovian-part">Markovian part</a><a id="Markovian-part-1"></a><a class="docs-heading-anchor-permalink" href="#Markovian-part" title="Permalink"></a></h3><p>To define the Markovian part of the model, we set the model by</p><pre><code class="language-julia hljs">using Catalyst, DiffEqJump
using Distributions, Random
using DelaySSAToolkit
rn = @reaction_network begin
    A, 0 --&gt; X
    B, Y --&gt; 0
end A B
u0 = [0,0] # X, Y
tf = 100.
tspan = (0,tf)
ps = [10., 1.] # A, B. The initial conditions are dummy variables, later on will be drawn from Gamma distributions </code></pre><p>where <span>$X$</span> is an extra species that represents the premature product in the delay channel that will turn into <span>$Y$</span> after a delay time <span>$\tau_n$</span> for each cell <span>$n$</span>. </p><h3 id="Non-Markovian-part"><a class="docs-heading-anchor" href="#Non-Markovian-part">Non-Markovian part</a><a id="Non-Markovian-part-1"></a><a class="docs-heading-anchor-permalink" href="#Non-Markovian-part" title="Permalink"></a></h3><pre><code class="language-julia hljs">τ = 1 # dummy variable, later on will be drawn from a Gamma distribution 
delay_trigger = Dict(1=&gt;[1=&gt;τ])
delay_complete = Dict(1=&gt;[1=&gt;-1, 2=&gt;1])
delay_interrupt = Dict()
jumpsys = convert(JumpSystem, rn, combinatoric_ratelaws=false)
dprob = DiscreteProblem(jumpsys,u0,tspan,ps)
de_chan0 = [[]]
delayjumpset = DelayJumpSet(delay_trigger, delay_complete, delay_interrupt)
jprob = DelayJumpProblem(jumpsys, dprob, DelayRejection(), delayjumpset, de_chan0, save_positions=(false,false))</code></pre><p>Here the value of <code>delay_trigger</code> is <code>[1=&gt;τ]</code>. This is of <code>Pair</code> type where the first index refers to the number of the delay channel, and the second index refers to the delay time. Such a definition is equivalent to appending a delay time <span>$\tau$</span> to the first delay channel provided that the first reaction <span>$A_n: 0\rightarrow X$</span> happened. <code>de_chan0</code> is the initial condition for delay channels, here we assume no on-going delay reactions.</p><h2 id="Heterogeneous-fixed-delays"><a class="docs-heading-anchor" href="#Heterogeneous-fixed-delays">Heterogeneous fixed delays</a><a id="Heterogeneous-fixed-delays-1"></a><a class="docs-heading-anchor-permalink" href="#Heterogeneous-fixed-delays" title="Permalink"></a></h2><h3 id="Define-EnsembleProblem"><a class="docs-heading-anchor" href="#Define-EnsembleProblem">Define <code>EnsembleProblem</code></a><a id="Define-EnsembleProblem-1"></a><a class="docs-heading-anchor-permalink" href="#Define-EnsembleProblem" title="Permalink"></a></h3><pre><code class="language-julia hljs"># the Gamma distribution in Julia uses shape α, scale θ, Gamma(α,θ). In the paper [1], Gamma distribution uses shape α and rate β. Thus, one needs to set the inverse.
function prob_func(prob, i ,repeat)
    rng = Random.seed!(i) # i is the index for each simulation
    A = rand(rng, Gamma(8,1/0.23)) 
    B = rand(rng, Gamma(9,1/625))
    τ = rand(rng, Gamma(7,1))
    remake(prob, p = [A,B], delay_trigger=Dict(1=&gt;[1=&gt;τ])) # update the new parameters
end
ensprob1 = EnsembleProblem(jprob, prob_func = prob_func)
@time ens1 = solve(ensprob1, SSAStepper(), EnsembleThreads(), trajectories = 40, saveat = 1.)</code></pre><p>For each simulation <span>$i$</span> (modelling an individual cell), a set of parameters <span>$A, B, τ$</span> is drawn from Gamma distributions, where we use the parameters defined as in [1]. One only needs to reconstruct the <code>DelayJumpProblem</code> by assigning new values to the parameter <code>p=[A,B]</code> and <code>delay_trigger =  Dict(1=&gt;[1=&gt;τ])</code> for each simulation. This can be easily done by invoking the <code>remake</code> function.</p><h2 id="Visualisation"><a class="docs-heading-anchor" href="#Visualisation">Visualisation</a><a id="Visualisation-1"></a><a class="docs-heading-anchor-permalink" href="#Visualisation" title="Permalink"></a></h2><p>We can plot the time evolution of 40 simulations. <img src="../../assets/heterogeneous_delay1.svg" alt="heterogeneous1"/></p><h2 id="Distributed-delays"><a class="docs-heading-anchor" href="#Distributed-delays">Distributed delays</a><a id="Distributed-delays-1"></a><a class="docs-heading-anchor-permalink" href="#Distributed-delays" title="Permalink"></a></h2><p>In [1, Section 3.2], the authors studied the case with distributed delays, where the delay can vary between reactions within and between cells. One can adopt the problem setting by changing few lines of code.</p><h3 id="Define-EnsembleProblem-2"><a class="docs-heading-anchor" href="#Define-EnsembleProblem-2">Define <code>EnsembleProblem</code></a><a class="docs-heading-anchor-permalink" href="#Define-EnsembleProblem-2" title="Permalink"></a></h3><pre><code class="language-julia hljs">function prob_func(prob, i ,repeat)
    rng = Random.seed!(i)
    α = rand(rng, Gamma(63,1/9))
    β = rand(rng, Gamma(10.,1/10.))
    A = rand(rng, Gamma(8,1/0.23))
    B = rand(rng, Gamma(9,1/625))
    τ = rand(rng, Gamma(α,1/β))
    remake(prob, p = [A,B], delay_trigger=Dict(1=&gt;[1=&gt;τ]))
end
ensprob2 = EnsembleProblem(jprob, prob_func = prob_func)</code></pre><h2 id="Visualisation-2"><a class="docs-heading-anchor" href="#Visualisation-2">Visualisation</a><a class="docs-heading-anchor-permalink" href="#Visualisation-2" title="Permalink"></a></h2><p>Note that a simulation of <span>$10^4$</span> samples with very high production number (upto ~1000 for X and ~3000 for Y) only takes few minutes on a laptop:</p><pre><code class="language-julia-repl hljs">julia&gt; @time ens2 = solve(ensprob2, SSAStepper(), EnsembleThreads(), trajectories = 10^4)

 78.925908 seconds (249.65 M allocations: 28.632 GiB, 6.60% gc time)
EnsembleSolution Solution of length 10000 with uType</code></pre><p>Here we plot the histogram of the number of unfinished reactant <span>$X$</span>s in the delay channel.</p><p><img src="../../assets/heterogeneous_delay2.svg" alt="heterogeneous2"/></p><h2 id="References"><a class="docs-heading-anchor" href="#References">References</a><a id="References-1"></a><a class="docs-heading-anchor-permalink" href="#References" title="Permalink"></a></h2><p>[1] Mark Jayson Cortez, Hyukpyo Hong, Boseung Choi, Jae Kyoung Kim, Krešimir Josić, &quot;Hierarchical Bayesian models of transcriptional and translational regulation processes with delays&quot;, Bioinformatics, 2021;, btab618, <a href="https://doi.org/10.1093/bioinformatics/btab618">https://doi.org/10.1093/bioinformatics/btab618</a></p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../delay_degradation/">« A birth-death example with delay degradation</a><a class="docs-footer-nextpage" href="../delay_oscillator/">An auto-regulatory model of oscillatory gene expression »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.17 on <span class="colophon-date" title="Friday 13 May 2022 07:03">Friday 13 May 2022</span>. Using Julia version 1.7.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
